source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,352 @@
+      return false;
+    }
+
+    // Reserve that many bytes in the string.
+    uint64_t size = index - pos;
+    value_.resize(size);
+    // Now copy from buffer to string.
+    data.copyOut(pos, index - pos, value_.data());
+    pos += (size + 1);
+    left -= (size + 1);
+
+    return true;
+  }
+
+  std::string toString() const { return absl::StrCat("[", value_, "]"); }
+
+private:
+  std::string value_;
+};
+
+// ByteN type is used as the last type in the Postgres message and contains
+// sequence of bytes. The length must be deduced from message length.
+class ByteN {
+public:
+  /**
+   * See above for parameter and return value description.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if (left > (data.length() - pos)) {
+      return false;
+    }
+    value_.resize(left);
+    data.copyOut(pos, left, value_.data());
+    pos += left;
+    left = 0;
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out = "[";
+    bool first = true;
+    for (const auto& i : value_) {
+      if (first) {
+        absl::StrAppend(&out, i);
+        first = false;
+      } else {
+        std::string buf;
+        buf = fmt::format(" {}", i);
+        absl::StrAppend(&out, buf);
+      }
+    }
+    absl::StrAppend(&out, "]");
+    return out;
+  }
+
+private:
+  std::vector<uint8_t> value_;
+};
+
+// VarByteN represents the structure consisting of 4 bytes of length
+// indicating how many bytes follow.
+// In Postgres documentation it is described as:
+// - Int32
+//   The number of bytes in the structure (this count does not include itself). Can be
+//   zero. As a special case, -1 indicates a NULL (no result). No value bytes follow in the NULL
+// case.
+//
+// - ByteN
+// The sequence of bytes representing the value. Bytes are present only when length has a positive
+// value.
+class VarByteN {
+public:
+  /**
+   * See above for parameter and return value description.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((left < sizeof(int32_t)) || ((data.length() - pos) < sizeof(int32_t))) {
+      return false;
+    }
+    len_ = data.peekBEInt<int32_t>(pos);
+    pos += sizeof(int32_t);
+    left -= sizeof(int32_t);
+    if (len_ < 1) {
+      // There is no payload if length is not positive.
+      value_.clear();
+      return true;
+    }
+    if ((left < static_cast<uint64_t>(len_)) ||
+        ((data.length() - pos) < static_cast<uint64_t>(len_))) {
+      return false;
+    }
+    value_.resize(len_);
+    data.copyOut(pos, len_, value_.data());
+    pos += len_;
+    left -= len_;
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out;
+    out = fmt::format("[({} bytes):", len_);
+
+    bool first = true;
+    for (const auto& i : value_) {
+      if (first) {
+        absl::StrAppend(&out, i);
+        first = false;
+      } else {
+        std::string buf;
+        buf = fmt::format(" {}", i);
```

Comment:
no need for format call.  Actually I think you can just replace this loop with `absl::StrAppend(&out, absl::StrJoin(value_, " "))`

Reply:
Thanks for this suggestion. It drastically simplified the logic (and reduced the loop with dreaded "first" to just single line!)